hudi-common/src/main/java/org/apache/hudi/common/table/read/IncrementalQueryAnalyzer.java

```diff
@@ -0,0 +1,425 @@
+    }
+  }
+
+  private HoodieTimeline getReadTimeline(HoodieTableMetaClient metaClient) {
+    HoodieTimeline timeline = metaClient.getCommitsAndCompactionTimeline().filterCompletedAndCompactionInstants();
+    return filterInstantsAsPerUserConfigs(metaClient, timeline, this.skipCompaction, this.skipClustering);
+  }
+
+  private HoodieTimeline getArchivedReadTimeline(HoodieTableMetaClient metaClient, String startInstant) {
+    HoodieArchivedTimeline archivedTimeline = metaClient.getArchivedTimeline(startInstant, false);
+    HoodieTimeline archivedCompleteTimeline = archivedTimeline.getCommitsTimeline().filterCompletedInstants();
+    return filterInstantsAsPerUserConfigs(metaClient, archivedCompleteTimeline, this.skipCompaction, this.skipClustering);
+  }
+
+  /**
+   * Filters out the unnecessary instants as per user specified configs.
+   *
+   * @param timeline The timeline.
+   *
+   * @return the filtered timeline
+   */
+  @VisibleForTesting
+  public static HoodieTimeline filterInstantsAsPerUserConfigs(HoodieTableMetaClient metaClient, HoodieTimeline timeline, boolean skipCompaction, boolean skipClustering) {
+    final HoodieTimeline oriTimeline = timeline;
+    if (metaClient.getTableType() == HoodieTableType.MERGE_ON_READ & skipCompaction) {
+      // the compaction commit uses 'commit' as action which is tricky
+      timeline = timeline.filter(instant -> !instant.getAction().equals(HoodieTimeline.COMMIT_ACTION));
+    }
+    if (skipClustering) {
+      timeline = timeline.filter(instant -> !ClusteringUtils.isClusteringInstant(instant, oriTimeline));
+    }
+    return timeline;
+  }
+
+  // -------------------------------------------------------------------------
+  //  Inner Class
+  // -------------------------------------------------------------------------
+  /**
+   * Builder for {@link IncrementalQueryAnalyzer}.
+   */
+  public static class Builder {
+    /**
+     * Start completion time.
+     */
+    private String startTime;
+    /**
+     * End completion time.
+     */
+    private String endTime;
+    private InstantRange.RangeType rangeType;
+    private HoodieTableMetaClient metaClient;
+    private boolean skipCompaction = false;
+    private boolean skipClustering = false;
+    /**
+     * Maximum number of instants to read per run.
+     */
+    private int limit = -1;
+
+    public Builder() {
+    }
+
+    public Builder startTime(String startTime) {
+      this.startTime = startTime;
+      return this;
+    }
+
+    public Builder endTime(String endTime) {
+      this.endTime = endTime;
+      return this;
+    }
+
+    public Builder rangeType(InstantRange.RangeType rangeType) {
+      this.rangeType = rangeType;
+      return this;
+    }
+
+    public Builder metaClient(HoodieTableMetaClient metaClient) {
+      this.metaClient = metaClient;
+      return this;
+    }
+
+    public Builder skipCompaction(boolean skipCompaction) {
```

Comment:
why are these needed? is it for table services and other internal processes?

Reply:
From a user standpoint, all of this should be transparent right.

Reply:
If we do not skip clustering or compaction for incremental/streaming read, there could be duplicates for downstream readers.

Reply:
I just add the interface here and the engine can decide whether to skip based on different use cases(batch or streaming).